checker/mod_checking.ml

```diff
@@ -89,40 +102,41 @@ and check_structure_field env mp lab res = function
       let c = Constant.make2 mp lab in
       check_constant_declaration env c cb
   | SFBmind mib ->
-      let kn = MutInd.make2 mp lab in
-      let kn = mind_of_delta res kn in
-      Indtypes.check_inductive env kn mib
+      let kn = KerName.make mp lab in
+      let kn = Mod_subst.mind_of_delta_kn res kn in
+      CheckInductive.check_inductive env kn mib
   | SFBmodule msb ->
       let () = check_module env (MPdot(mp,lab)) msb in
       Modops.add_module msb env
   | SFBmodtype mty ->
       check_module_type env mty;
-      add_modtype (MPdot(mp,lab)) mty env
+      add_modtype mty env
 
 and check_mexpr env mse mp_mse res = match mse with
   | MEident mp ->
       let mb = lookup_module mp env in
-      (subst_and_strengthen mb mp_mse).mod_type
+      (Modops.strengthen_and_subst_mb mb mp_mse false).mod_type
   | MEapply (f,mp) ->
       let sign = check_mexpr env f mp_mse res in
-      let farg_id, farg_b, fbody_b = destr_functor sign in
-      let mtb = module_type_of_module (Some mp) (lookup_module mp env) in
-      check_subtypes env mtb farg_b;
-      subst_signature (map_mbid farg_id mp) fbody_b
-  | MEwith _ -> error_with_module ()
+      let farg_id, farg_b, fbody_b = Modops.destr_functor sign in
+      let mtb = Modops.module_type_of_module (lookup_module mp env) in
+      let _ = check_subtypes env mtb farg_b in (* Should we stop inferring constraints? *)
```

Comment:
Shouldn't we check that the constraints are satisfied instead of dropping them?

Reply:
Right, what function should I call?

Reply:
Although, isn't it already checked by the kernel at subtyping time? If not, when does the kernel check it?

Reply:
The kernel adds the constraints to the environment. This part corresponds to some subcall of translate_module so the constraints are first added to the module_body that function produces, then when the module is added to the safe env the constraints are added.
When checking use Environ.check_constraints (returns bool).